### PR TITLE
modified prepare file with cut off energy convergence

### PR DESCRIPTION
**Pull request by KaworuIkari**

### CODE REVIEW

File: scripts/copper/bulk/prepare.py [hunk 1]
```diff
@@ -16,19 +16,32 @@
...
     for n_k_points in range(1, 11):
-        config = CastepConfig(data_path / f"bulk_{n_k_points}", "bulk")
+        config = CastepConfig(data_path_k / f"bulk_{n_k_points}", "bulk")
         params = BulkOptimizationParams(n_k_points=n_k_points)
         calculator = get_bulk_optimization_calculator(bulk_copper, params, config)
         prepare_calculator_with_submit_script(calculator)
 
         calculators.append(calculator)
+    # cut off energy convergence test
+    for ce in range(300, 750, 50):
```
Comment (Matt-Ord) [inline scripts/copper/bulk/prepare.py:36]: rename ce to cutoff_energy to be clearer

File: scripts/copper/bulk/prepare.py [hunk 2]
```diff
@@ -16,19 +16,32 @@
...
+    # k points convergence test
     for n_k_points in range(1, 11):
-        config = CastepConfig(data_path / f"bulk_{n_k_points}", "bulk")
+        config = CastepConfig(data_path_k / f"bulk_{n_k_points}", "bulk")
         params = BulkOptimizationParams(n_k_points=n_k_points)
         calculator = get_bulk_optimization_calculator(bulk_copper, params, config)
         prepare_calculator_with_submit_script(calculator)
 
         calculators.append(calculator)
+    # cut off energy convergence test
```
Comment (Matt-Ord) [inline scripts/copper/bulk/prepare.py:35]: maybe extract each of these into a function, either in ppc library or just in this file

File: scripts/copper/bulk/prepare.py [hunk 3]
```diff
@@ -16,19 +16,32 @@
...
+    for ce in range(300, 750, 50):
+        config = CastepConfig(data_path_ce / f"bulk_{ce}", "bulk")
+        params = BulkOptimizationParams(n_k_points=11, cut_off_energy=ce)
+        calculator = get_bulk_optimization_calculator(bulk_copper, params, config)
+        prepare_calculator_with_submit_script(calculator)
+
+    prepare_submit_all_script(calculators, data_path_k)
+
+    prepare_submit_all_script(calculators, data_path_ce)
+    copy_files_to_hpc(data_path_k, PosixPath("copper/bulk"))
```
Comment (Matt-Ord) [inline scripts/copper/bulk/prepare.py:45]: move this up above all of the cutoff energy code

File: scripts/copper/bulk/prepare.py [hunk 4]
```diff
@@ -16,19 +16,32 @@
 
 if __name__ == "__main__":
     bulk_copper = bulk("Cu", "fcc", 3.8)  #
-    data_path = Path("data/copper/bulk")
+    data_path_k = Path("data/copper/bulk_k")
+    data_path_ce = Path("data/copper/bulk_ce")
```
Comment (Matt-Ord) [inline scripts/copper/bulk/prepare.py:20]: only define this when it is needed (ie below k convergene code. either just call both of these varibles data_path or call it by the full name data_path_cutoff_energy. Only use acronyms when they are `standard` (it is not clear to me that ce stands for cutoff_energy)

File: scripts/copper/bulk/prepare.py [hunk 5]
```diff
@@ -16,19 +16,32 @@
 
 if __name__ == "__main__":
     bulk_copper = bulk("Cu", "fcc", 3.8)  #
```
Comment (Matt-Ord) [inline scripts/copper/bulk/prepare.py:18]: remove # here

File: scripts/copper/bulk/prepare.py [hunk 6]
```diff
@@ -16,19 +16,32 @@
...
-    data_path = Path("data/copper/bulk")
+    data_path_k = Path("data/copper/bulk_k")
+    data_path_ce = Path("data/copper/bulk_ce")
+    prepare_clean_directory(data_path_k)
 
-    prepare_clean_directory(data_path)
+    prepare_clean_directory(data_path_ce)
```
Comment (Matt-Ord) [inline scripts/copper/bulk/prepare.py:23]: only prepare when needed (ie below kpoints code)

File: scripts/copper/bulk/prepare.py (hunk 2 above)
Comment (Matt-Ord) [inline scripts/copper/bulk/prepare.py:35]: is this comment helpful?

**Review by Matt-Ord — CHANGES_REQUESTED** [review]
Great work! left some comments

File: scripts/copper/bulk/prepare.py [hunk 7]
```diff
@@ -16,19 +16,32 @@
...
+    # cut off energy convergence test
+    for ce in range(300, 750, 50):
+        config = CastepConfig(data_path_ce / f"bulk_{ce}", "bulk")
+        params = BulkOptimizationParams(n_k_points=11, cut_off_energy=ce)
+        calculator = get_bulk_optimization_calculator(bulk_copper, params, config)
+        prepare_calculator_with_submit_script(calculator)
+
+    prepare_submit_all_script(calculators, data_path_k)
+
+    prepare_submit_all_script(calculators, data_path_ce)
```
[the remark this answers is not in the document]
Comment (KaworuIkari) [reply scripts/copper/bulk/prepare.py:44]: calculators = list[Castep]()
    
    what does this line of code do in line 25

**Review by Matt-Ord — APPROVED** [review]

**Comment by Matt-Ord** [comment]
Brilliant work - well done!

**Pull request merged**